tagmanager/src/tm_parser.c

```diff
@@ -0,0 +1,647 @@
+	MAP_ENTRY(html_map),
+	MAP_ENTRY(f77_map),
+	MAP_ENTRY(c_map),	/* GLSL - same as C */
+	MAP_ENTRY(matlab_map),
+	MAP_ENTRY(vala_map),
+	MAP_ENTRY(actionscript_map),
+	MAP_ENTRY(nsis_map),
+	MAP_ENTRY(markdown_map),
+	MAP_ENTRY(txt2tags_map),
+	MAP_ENTRY(abc_map),
+	MAP_ENTRY(verilog_map),
+	MAP_ENTRY(r_map),
+	MAP_ENTRY(cobol_map),
+	MAP_ENTRY(objc_map),
+	MAP_ENTRY(asciidoc_map),
+	MAP_ENTRY(abaqus_map),
+	MAP_ENTRY(rust_map),
+	MAP_ENTRY(go_map),
+	MAP_ENTRY(json_map),
+	MAP_ENTRY(php_map),	/* Zephir - same as PHP */
+	MAP_ENTRY(powershell_map),
+};
+
+
+TMTagType tm_parser_get_tag_type(gchar kind, TMParserType lang)
+{
+	TMParserMap *map = &parser_map[lang];
+	guint i;
+
+	for (i = 0; i < map->size; i++)
+	{
+		TMParserMapEntry *entry = &map->entries[i];
+
+		if (entry->kind == kind)
+			return entry->type;
+	}
+	return tm_tag_undef_t;
+}
+
+
+gchar tm_parser_get_tag_kind(TMTagType type, TMParserType lang)
+{
+	TMParserMap *map = &parser_map[lang];
+	guint i;
+
+	for (i = 0; i < map->size; i++)
+	{
+		TMParserMapEntry *entry = &map->entries[i];
+
+		if (entry->type == type)
+			return entry->kind;
+	}
+	return '-';
+}
+
+
+void tm_parser_verify_type_mappings(void)
+{
+	gsize parser_map_size = sizeof(parser_map) / sizeof(TMParserMap);
+	TMParserType lang;
+
+	if (TM_PARSER_COUNT > tm_ctags_get_lang_count())
+	{
+		g_warning("More parsers defined in Geany than in ctags");
+		return;
+	}
+
+	if (parser_map_size != TM_PARSER_COUNT)
+	{
+		g_warning("Different number of parsers and tag type mappings");
+		return;
+	}
+
+	for (lang = 0; lang < TM_PARSER_COUNT; lang++)
+	{
+		const gchar *kinds = tm_ctags_get_lang_kinds(lang);
+		TMParserMap *map = &parser_map[lang];
+		gchar presence_map[256];
+		guint i;
+
+		/* TODO: check also regex parser mappings. At the moment there's no way
+		 * to access regex parser definitions in ctags */
+		if (tm_ctags_is_using_regex_parser(lang))
+			continue;
+
+		if (map->size != strlen(kinds))
+		{
+			g_warning("Different number of tag types in TM (%d) and ctags (%d) for %s",
+				map->size, (int)strlen(kinds), tm_ctags_get_lang_name(lang));
+			continue;
+		}
+
+		memset(presence_map, 0, sizeof(presence_map));
+		for (i = 0; i < map->size; i++)
+		{
+			gboolean ctags_found = FALSE;
+			gboolean tm_found = FALSE;
+			guint j;
+
+			for (j = 0; j < map->size; j++)
+			{
+				/* check that for every type in TM there's a type in ctags */
+				if (map->entries[i].kind == kinds[j])
+					ctags_found = TRUE;
+				/* check that for every type in ctags there's a type in TM */
+				if (map->entries[j].kind == kinds[i])
+					tm_found = TRUE;
+				if (ctags_found && tm_found)
+					break;
+			}
+			if (!ctags_found)
+				g_warning("Tag type '%c' found in TM but not in ctags for %s",
+					map->entries[i].kind, tm_ctags_get_lang_name(lang));
+			if (!tm_found)
+				g_warning("Tag type '%c' found in ctags but not in TM for %s",
+					kinds[i], tm_ctags_get_lang_name(lang));
+
+			presence_map[map->entries[i].kind]++;
```

Comment:
```gcc
../../../tagmanager/src/tm_parser.c: In function 'tm_parser_verify_type_mappings':
../../../tagmanager/src/tm_parser.c:637:16: warning: array subscript has type 'char' [-Wchar-subscripts]
    presence_map[map->entries[i].kind]++;
                ^
```

casting to `unsigned char` should be fine.

---
BTW, `presence_map` could be 127 bytes only and indexed as `((unsigned char) map->entries[i].kind) & 0x0f`, kind names won't use non-ASCII letters.  But that's probably an over-optimization that might blow off on us someday.